### DISCUSSION TITLE
Best practices for chaining transactions

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/v2-periphery#3] [Ideas]
During our last conversations we touched upon the need for chains of transactions. Some examples would be:

- wrap/unwrap utility assets (e.g. ETH/WETH) + deposit/withdraw
- wrap/unwrap vault assets (e.g. you deposit cUSDC but we wrap it in vault-cUSDC) + deposit/withdraw
- any other combination of at least 2 atomic actions

The requirements for these chains to exist might affect (1) the methods we need to implement (see sablierhq/v2-core#39 - things like `onBehalfOf` so that a contract can interact on behalf of a sender) and (2) the logic of our data structures.

I'm opening this discussion to reference any interesting architecture or methodology that might help us achieve these v2-periphery functionality and see what updates we need to make to v2-core as a result.

**Comment by razgraf** [comment]
Some products we should explore:

- [Uniswap](https://github.com/uniswap/) (doing wrapping/unwrapping of utility assets and trades in the same tx)
- [Furucombo](https://docs.furucombo.app/) (the king of transaction chaining)
- Yearn (their strategies usually involve a lot of atomic actions, although most are done for one user - the vault)

**Comment by PaulRBerg** [comment]
> to reference any interesting architecture or methodology.

Oh, I have an idea! It's called [PRBProxy](https://github.com/paulrberg/prb-proxy) 😅

What we could do is:

1. If the user wants to create only one stream (a simple one), we can skip using the proxy and create the stream directly from the EOA.
2. For anything more complicated (batching streams, wrapping native assets, vault tokens à la ERC-4626), we could use PRBProxy to compose the required atomic transactions.

**Comment by PaulRBerg** [comment]
Locking in favor of https://github.com/sablierhq/v2-periphery/discussions/11.
